mlflow/gateway/providers/openai.py

```diff
@@ -121,18 +150,200 @@ async def chat_stream(
+        warehouse_id = os.environ.get("DATABRICKS_WAREHOUSE_ID")
+        if warehouse_id is None:
+            raise HTTPException(
+                status_code=400,
+                detail="DATABRICKS_WAREHOUSE_ID environment variable is not set",
+            )
+
+        from fastapi.encoders import jsonable_encoder
+
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+
+        token_usage_accumulator = TokenUsageAccumulator()
+        user_tool_messages = [m for m in payload["messages"] if m["role"] == "tool"]
+        user_tool_calls = next(
+            (m["tool_calls"] for m in payload["messages"] if "tool_calls" in m), None
+        )
+        if (
+            user_tool_messages
+            and user_tool_calls
+            and (result := parse_uc_functions(payload["messages"][0]["content"]))
+        ):
+            uc_func_calls, uc_func_messages = result
+            messages = [
+                *[m for m in payload["messages"] if m["role"] == "tool" or "tool_calls" in m],
+                # Join UC function calls and user tool calls
+                {
+                    "role": "assistant",
+                    "content": None,
+                    "tool_calls": uc_func_calls + user_tool_calls,
+                },
+                *uc_func_messages,
+                *user_tool_messages,
+            ]
+            resp = await send_request(
+                headers=self._request_headers,
+                base_url=self._request_base_url,
+                path="chat/completions",
+                payload=self._add_model_to_payload_if_necessary(
+                    {
+                        **payload,
+                        "messages": messages,
+                    }
+                ),
+            )
+            token_usage_accumulator.update(resp.get("usage", {}))
+        elif any(t["type"] == _UC_FUNCTION for t in payload.get("tools", [])):
+            updated_tools = []
+            uc_func_mapping: Dict[str, "FunctionInfo"] = {}
+            for tool in payload.get("tools", []):
+                if tool["type"] == _UC_FUNCTION:
+                    function_name = tool[_UC_FUNCTION]["name"]
+                    function = workspace_client.functions.get(function_name)
+                    param_metadata = get_func_schema(function)
+                    t = {
+                        "type": "function",
+                        "function": param_metadata,
+                    }
+                    uc_func_mapping[t["function"]["name"]] = function
+                    updated_tools.append(t)
+                else:
+                    updated_tools.append(tool)
+
+            payload["tools"] = updated_tools
+
+            messages = payload.pop("messages", [])
+            uc_func_calls = []
+            user_tool_calls = []
+            resp = None
+            for _ in range(20):  # loop until we get a response without tool_calls
```

Comment:
@mengxr do you have an example that requires multiple iterations until we get a response without tool_calls?